src/NLog/Contexts/GlobalContext.cs

```diff
@@ -0,0 +1,220 @@
+        private Dictionary<string, object> dict;
+        private HashSet<String> keys;
+
+        private GlobalContext()
+        {
+            this.dict = new Dictionary<string, object>();
+            keys = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Current instance of the context.
+        /// </summary>
+        public static IContext Instance
+        {
+            get
+            {
+                if (currentInstance == null)
+                {
+                    lock (syncRoot)
+                    {
+                        if (currentInstance == null)
+                        {
+                            currentInstance = new GlobalContext();
+                        }
+                    }
+                }
+                return currentInstance;
+            }
+        }
+
+        /// <summary>
+        /// Returns the keys in the context.
+        /// </summary>
+        public HashSet<string> Keys
+        {
+            get
+            {
+                return keys;
+            }
+        }
+
+        /// <summary>
+        /// Set / Get the key in the context.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns>The key value, if defined; otherwise <c>null</c>.</returns>
+        public object this[string key]
+        {
+            get
+            {
+                return this.TryGet(key);
+            }
+            set
+            {
+                lock (syncRoot)
+                {
+                    this.dict[key] = value;
+                    if (!keys.Contains(key))
+                        keys.Add(key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clears the content.
+        /// </summary>
+        public void Clear()
+        {
+            lock (syncRoot)
+            {
+                this.dict.Clear();
+                this.keys.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the specified key exists in the Context.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        /// <returns>A boolean indicating whether the specified key exists in current context.</returns>
+        public bool Contains(string key)
+        {
+            lock (syncRoot)
+            {
+                return this.dict.ContainsKey(key);
+            }
+        }
+
+        /// <summary>
+        /// Removes the specified key from the Context.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        public void Remove(string key)
+        {
+            lock (syncRoot)
+            {
+                this.dict.Remove(key);
+            }
+        }
+
+        /// <summary>
+        /// Sets an key in the context.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        public void Set(string key, object value)
```

Comment:
there is not "get", so I find a "set" a bit strange. I think we should remove it.